<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Obtain risk scores and alleles &mdash; hipsnp  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/sg_gallery.css" type="text/css" />
      <link rel="stylesheet" href="../_static/sg_gallery-binder.css" type="text/css" />
      <link rel="stylesheet" href="../_static/sg_gallery-dataframe.css" type="text/css" />
      <link rel="stylesheet" href="../_static/sg_gallery-rendered-html.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Reference" href="../api.html" />
    <link rel="prev" title="Examples" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> hipsnp
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installing</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Examples</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Obtain risk scores and alleles</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../maintaining.html">Maintaining</a></li>
<li class="toctree-l1"><a class="reference internal" href="../whats_new.html">What’s new</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">hipsnp</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="index.html">Examples</a> &raquo;</li>
      <li>Obtain risk scores and alleles</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/auto_examples/plot_example_mock_data.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="sphx-glr-download-link-note admonition note">
<p class="admonition-title">Note</p>
<p>Click <a class="reference internal" href="#sphx-glr-download-auto-examples-plot-example-mock-data-py"><span class="std std-ref">here</span></a>
to download the full example code</p>
</div>
<section class="sphx-glr-example-title" id="obtain-risk-scores-and-alleles">
<span id="sphx-glr-auto-examples-plot-example-mock-data-py"></span><h1>Obtain risk scores and alleles<a class="headerlink" href="#obtain-risk-scores-and-alleles" title="Permalink to this headline"></a></h1>
<p>This example uses a mock dataset file with genotic information hosted in GIN.
The alleles are extracted from the BGEN files and a risk score is calculated</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># # Authors: Oscar Portoles &lt;o.portoles@fz-juelich.de&gt;</span>
<span class="c1">#            Federico Raimondo &lt;f.raimondo@fz-juelich.de&gt;</span>
<span class="c1">#</span>
<span class="c1"># License: AGPL</span>

<span class="kn">from</span> <span class="nn">tempfile</span> <span class="kn">import</span> <a href="https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp" title="tempfile.mkdtemp" class="sphx-glr-backref-module-tempfile sphx-glr-backref-type-py-function"><span class="n">mkdtemp</span></a>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">from</span> <span class="nn">hipsnp.hipsnp</span> <span class="kn">import</span> <span class="n">genotype_from_datalad</span>
<span class="kn">from</span> <span class="nn">hipsnp.utils</span> <span class="kn">import</span> <span class="n">configure_logging</span>
</pre></div>
</div>
<p>Set the logging level to info to see extra information</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">configure_logging</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="s1">&#39;INFO&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p class="sphx-glr-script-out">Out:</p>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>2021-12-17 09:25:53,878 - HIPSNP - INFO - ===== Lib Versions =====
2021-12-17 09:25:53,879 - HIPSNP - INFO - numpy: 1.21.4
2021-12-17 09:25:53,879 - HIPSNP - INFO - scipy: 1.7.3
2021-12-17 09:25:53,879 - HIPSNP - INFO - bgen_reader: 4.0.8
2021-12-17 09:25:53,879 - HIPSNP - INFO - datalad: 0.15.4
2021-12-17 09:25:53,879 - HIPSNP - INFO - pandas: 1.3.5
2021-12-17 09:25:53,879 - HIPSNP - INFO - ========================
</pre></div>
</div>
<p>Get a temporary directory to work. Use another directory if you want to
keep the data.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><a href="https://docs.python.org/3/library/stdtypes.html#str" title="builtins.str" class="sphx-glr-backref-module-builtins sphx-glr-backref-type-py-class sphx-glr-backref-instance"><span class="n">workdir</span></a> <span class="o">=</span> <a href="https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp" title="tempfile.mkdtemp" class="sphx-glr-backref-module-tempfile sphx-glr-backref-type-py-function"><span class="n">mkdtemp</span></a><span class="p">()</span>
</pre></div>
</div>
<p>Obtain a genotype from a datalad dataset. We can use the same directory
as work and data directory.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><a href="https://docs.python.org/3/library/stdtypes.html#str" title="builtins.str" class="sphx-glr-backref-module-builtins sphx-glr-backref-type-py-class sphx-glr-backref-instance"><span class="n">source</span></a> <span class="o">=</span> <span class="s1">&#39;https://gin.g-node.org/juaml/datalad-example-bgen&#39;</span>
<span class="n">genotype</span> <span class="o">=</span> <span class="n">genotype_from_datalad</span><span class="p">(</span>
    <span class="n">rsids</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;RSID_1&#39;</span><span class="p">,</span> <span class="s1">&#39;RSID_2&#39;</span><span class="p">,</span> <span class="s1">&#39;RSID_3&#39;</span><span class="p">],</span> <span class="n">chromosomes</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">],</span>
    <span class="n">datalad_source</span><span class="o">=</span><a href="https://docs.python.org/3/library/stdtypes.html#str" title="builtins.str" class="sphx-glr-backref-module-builtins sphx-glr-backref-type-py-class sphx-glr-backref-instance"><span class="n">source</span></a><span class="p">,</span> <a href="https://docs.python.org/3/library/stdtypes.html#str" title="builtins.str" class="sphx-glr-backref-module-builtins sphx-glr-backref-type-py-class sphx-glr-backref-instance"><span class="n">workdir</span></a><span class="o">=</span><a href="https://docs.python.org/3/library/stdtypes.html#str" title="builtins.str" class="sphx-glr-backref-module-builtins sphx-glr-backref-type-py-class sphx-glr-backref-instance"><span class="n">workdir</span></a><span class="p">,</span> <span class="n">datadir</span><span class="o">=</span><a href="https://docs.python.org/3/library/stdtypes.html#str" title="builtins.str" class="sphx-glr-backref-module-builtins sphx-glr-backref-type-py-class sphx-glr-backref-instance"><span class="n">workdir</span></a><span class="p">,)</span>
</pre></div>
</div>
<p class="sphx-glr-script-out">Out:</p>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>2021-12-17 09:25:54,011 - HIPSNP - INFO - Chromosomes needed: [&#39;1&#39;]
2021-12-17 09:25:54,012 - HIPSNP - INFO - Getting chromosome 1 for RSID(s) [&#39;RSID_1&#39; &#39;RSID_2&#39; &#39;RSID_3&#39;]
[INFO] Cloning dataset to Dataset(/tmp/tmp86z2yp5v)
[INFO] Attempting to clone from https://gin.g-node.org/juaml/datalad-example-bgen to /tmp/tmp86z2yp5v
[INFO] Start enumerating objects
[INFO] Start counting objects
[INFO] Start compressing objects
[INFO] Completed clone attempts for Dataset(/tmp/tmp86z2yp5v)
[INFO] access to 1 dataset sibling github-lfs not auto-enabled, enable with:
|               datalad siblings -d &quot;/tmp/tmp86z2yp5v&quot; enable -s github-lfs
2021-12-17 09:25:57,420 - HIPSNP - INFO - Getting files: [PosixPath(&#39;/tmp/tmp86z2yp5v/imputation/example_c1_v0.sample&#39;), PosixPath(&#39;/tmp/tmp86z2yp5v/imputation/example_c1_v0.bgen&#39;)]
2021-12-17 09:26:00,052 - HIPSNP - INFO - Converting to BGEN: qctool -g /tmp/tmp86z2yp5v/imputation/example_c1_v0.bgen -s /tmp/tmp86z2yp5v/imputation/example_c1_v0.sample -incl-rsids /tmp/tmp86z2yp5v/rsids_chromosome1.txt -og /tmp/tmp86z2yp5v/chromosome1.bgen -ofiletype bgen_v1.2 -bgen-bits 8

2021-12-17 09:26:03,447 - HIPSNP - INFO - Reading 1 bgen files...
2021-12-17 09:26:03,448 - HIPSNP - INFO - Reading /tmp/tmp86z2yp5v/chromosome1.bgen
</pre></div>
</div>
<p>Now we are ready to obtain the alleles of each rsid and sample in the data</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gen_allele</span><span class="p">,</span> <span class="n">gen_012</span> <span class="o">=</span> <span class="n">genotype</span><span class="o">.</span><span class="n">alleles</span><span class="p">()</span>

<span class="c1"># For example, we can count the number of times that an allele appears</span>
<span class="n">gen_allele</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;RSID_2&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
<p class="sphx-glr-script-out">Out:</p>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>2021-12-17 09:26:03,464 - HIPSNP - INFO - Samples are not consolidated across RSIDs. Consolidating...
2021-12-17 09:26:03,465 - HIPSNP - INFO - Calculating genotypes for 2 SNPs and                     500 samples ...

GG    341
AG    147
AA     12
Name: RSID_2, dtype: int64
</pre></div>
</div>
<p>To compute a poligenetic risk score we need a file or a pandas dataframe with
the weights associated to each allele and RSID. We can retrieve one from
<a class="reference external" href="https://www.pgscatalog.org/">https://www.pgscatalog.org/</a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># In this case we use a weights file generated for this mock datasets.</span>
<a href="https://docs.python.org/3/library/stdtypes.html#str" title="builtins.str" class="sphx-glr-backref-module-builtins sphx-glr-backref-type-py-class sphx-glr-backref-instance"><span class="n">path_to_weights</span></a> <span class="o">=</span> <span class="s1">&#39;./data/weights_all.csv&#39;</span>

<span class="c1"># Now we can obtain the risk score for each rsids given the samples on the</span>
<span class="c1"># dataset and the dosage (amount of the effect allele) of each rsids and sample</span>
<span class="n">dosage</span><span class="p">,</span> <span class="n">risk</span> <span class="o">=</span> <span class="n">genotype</span><span class="o">.</span><span class="n">riskscores</span><span class="p">(</span><span class="n">weights</span><span class="o">=</span><a href="https://docs.python.org/3/library/stdtypes.html#str" title="builtins.str" class="sphx-glr-backref-module-builtins sphx-glr-backref-type-py-class sphx-glr-backref-instance"><span class="n">path_to_weights</span></a><span class="p">)</span>
</pre></div>
</div>
<p class="sphx-glr-script-out">Out:</p>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>2021-12-17 09:26:03,488 - HIPSNP - INFO - Samples are not consolidated across RSIDs. Consolidating...
2021-12-17 09:26:03,490 - HIPSNP - INFO - Calculating riskscores for 2 SNPs and                     500 samples ...
</pre></div>
</div>
<p>Then, for example, we can visualize the risk score of each sample</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">risk</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ls</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;*&#39;</span><span class="p">,</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<img src="../_images/sphx_glr_plot_example_mock_data_001.png" srcset="../_images/sphx_glr_plot_example_mock_data_001.png" alt="plot example mock data" class = "sphx-glr-single-img"/><p class="sphx-glr-script-out">Out:</p>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>&lt;AxesSubplot:&gt;
</pre></div>
</div>
<p>Since we used a temporary directory, we need to delete it</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><a href="https://docs.python.org/3/library/shutil.html#shutil.rmtree" title="shutil.rmtree" class="sphx-glr-backref-module-shutil sphx-glr-backref-type-py-function"><span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span></a><span class="p">(</span><a href="https://docs.python.org/3/library/stdtypes.html#str" title="builtins.str" class="sphx-glr-backref-module-builtins sphx-glr-backref-type-py-class sphx-glr-backref-instance"><span class="n">workdir</span></a><span class="p">)</span>
</pre></div>
</div>
<p class="sphx-glr-timing"><strong>Total running time of the script:</strong> ( 0 minutes  11.035 seconds)</p>
<div class="sphx-glr-footer class sphx-glr-footer-example docutils container" id="sphx-glr-download-auto-examples-plot-example-mock-data-py">
<div class="sphx-glr-download sphx-glr-download-python docutils container">
<p><a class="reference download internal" download="" href="../_downloads/676865f9a5f5d9ec02deef362b2ee8d5/plot_example_mock_data.py"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Python</span> <span class="pre">source</span> <span class="pre">code:</span> <span class="pre">plot_example_mock_data.py</span></code></a></p>
</div>
<div class="sphx-glr-download sphx-glr-download-jupyter docutils container">
<p><a class="reference download internal" download="" href="../_downloads/87e5e12f81edf214c1242c177c9c5c49/plot_example_mock_data.ipynb"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Jupyter</span> <span class="pre">notebook:</span> <span class="pre">plot_example_mock_data.ipynb</span></code></a></p>
</div>
</div>
<p class="sphx-glr-signature"><a class="reference external" href="https://sphinx-gallery.github.io">Gallery generated by Sphinx-Gallery</a></p>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="Examples" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../api.html" class="btn btn-neutral float-right" title="Reference" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Authors of hipsnp.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  
<div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    <span class="fa fa-book"> Other Versions</span>
    v: main
    <span class="fa fa-caret-down"></span>
  </span>
  <div class="rst-other-versions">
    <dl>
      <dt>Tags</dt>
      <dd><a href="../../v0.0.1/index.html">v0.0.1</a></dd>
    </dl>
    <dl>
      <dt>Branches</dt>
      <dd><a href="plot_example_mock_data.html">main</a></dd>
    </dl>
  </div>
</div><script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>